pennylane/gradients/hessian_transform.py

```diff
@@ -0,0 +1,223 @@
+      evaluated ``hessian_tapes``. It should accept a list of numeric results with
+      length ``len(hessian_tapes)``, and return the Hessian matrix.
+
+    Once defined, the quantum Hessian transform can be used as follows:
+
+    >>> hessian_tapes, processing_fn = my_custom_hessian(tape, *hessian_kwargs)
+    >>> res = execute(tapes, dev, interface="autograd", gradient_fn=qml.gradients.param_shift)
+    >>> jacobian = processing_fn(res)
+
+    Alternatively, Hessian transforms can be applied directly to QNodes, in which case
+    the execution is implicit:
+
+    >>> fn = my_custom_gradient(qnode, *gradient_kwargs)
+    >>> fn(weights)  # transformed function takes the same arguments as the QNode
+    1.2629730888100839
+
+    .. note::
+
+        The input tape might have parameters of various types, including NumPy arrays,
+        JAX DeviceArrays, and TensorFlow and PyTorch tensors.
+
+        If the Hessian transform is written in a autodiff-compatible manner, either by
+        using a framework such as Autograd or TensorFlow, or by using ``qml.math`` for
+        tensor manipulation, then higher-order derivatives will also be supported.
+
+        Alternatively, you may use the ``tape.unwrap()`` context manager to temporarily
+        convert all tape parameters to NumPy arrays and floats:
+
+        >>> with tape.unwrap():
+        ...     params = tape.get_parameters()  # list of floats
+    """
+
+    def __init__(
+        self, transform_fn, expand_fn=expand_invalid_trainable, differentiable=True, hybrid=True
+    ):
+        self.hybrid = hybrid
+        super().__init__(transform_fn, expand_fn=expand_fn, differentiable=differentiable)
+
+    @staticmethod
+    def _jacobian_trainable_args(args, interface):
+        """Return the indices of QNode arguments for which a Jacobian was
+        computed by `qml.transforms.classical_jacobian` with argnum=None.
+        """
+        trainable_args = []
+
+        if interface == "autograd":
+            for idx, arg in enumerate(args):
+                # TODO: make default False once this change is done in qml.jacobian
+                if getattr(arg, "requires_grad", True):
+                    trainable_args.append(idx)
+
+        elif interface == "jax":
+            trainable_args = [0]
+
+        # Torch and Tensorflow interfaces are not considered since `classical_jacobian`
+        # always returns a tuple for them, thus not invoking this function.
+
+        return trainable_args
+
+    def default_qnode_wrapper(self, qnode, targs, tkwargs):
+        # Here, we overwrite the QNode execution wrapper in order
+        # to take into account that classical processing may be present
+        # inside the QNode.
+        hybrid = tkwargs.pop("hybrid", self.hybrid)
+        _wrapper = super().default_qnode_wrapper(qnode, targs, tkwargs)
+        cjac_fn = qml.transforms.classical_jacobian(qnode, expand_fn=expand_invalid_trainable)
```

Comment:
Good question! This line is directly taken from the `gradient_transform`, maybe @josh146 would know?
Personally, I don't see why `self.expand_fn` wasn't used here.

Reply:
Can `self.expand_fn` differ from `expand_invalid_trainable`? That is probably the reason why it is hardcoded in the original `gradient_transform` 🤔

Reply:
Well it is a keyword argument to the transform 🤔

Reply:
@dwierichs it looks like you hardcoded it in [this PR](https://github.com/PennyLaneAI/pennylane/pull/1734/files?authenticity_token=UffkpQLOJTdmEoCchuZ120GVxlRZtA2cybl6UThCdWwyROotlx3OnrY4iyKdgG4rDSbFt24JiG4WHIlKihxjfQ&file-filters%5B%5D=.py#diff-7498bddbc39a804610097ec71b60020e226e1f244f3d4d124faa6be9cfdd1c91) 😄

Reply:
Oh gosh :facepalm:

Reply:
I don't see a reason why this should be hardcoded. If a `gradient_transform` is passed a custom `expand_fn`, the classical Jacobian might not even have matching dimensions, right?

Reply:
Maybe so that the batch transform can use a different expansion function (as set from the keyword argument) than the gradient transform (always `expand_invalid_trainable`)? Not sure if that scenario makes any sense.

Reply:
Indeed, the following code crashes, as the custom expand function avoids any expansion from happening, so that the quantum Jacobian has a single entry.
Meanwhile the classical Jacobian expands the `Rot` operation into three, yielding a 3x3 matrix as classical Jacobian that is incompatible with the single-element quantum Jacobian.
```python
expand_fn = qml.transforms.create_expand_fn(1, stop_at=qml.BooleanFn(lambda op: True))

@partial(gradient_transform, expand_fn=expand_fn)
def eye(tape):
    proc = lambda res: res
    return [tape], proc

dev = qml.device("default.qubit", wires=2)

def ansatz(x):
    qml.Rot(x[0], x[1]/2, x[2], wires=0)

@qml.qnode(dev)
def qnode(x):
    ansatz(x)
    return qml.expval(qml.PauliX(0))

eye(qnode)(pnp.array([0.4, 0.3, 0.1]))
```

Reply:
(Note that the modification of at least one of the parameters is necessary because the mismatching shapes are not noticed by PL whenever the classical Jacobian is the identity :D)

Reply:
Ah yes I see what you are saying now :) The classical Jacobian should definitely be using the same expansion function then (`self.expand_fn`). I'll update this here for the Hessian transform, but we should do the same for the gradient transform as well.